### PR TITLE
Fix orphaned spans in WCF traces

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

Fixes orphan traces in WCF

## Reason for change

Many customers have reported that in v3, with "delayed WCF" instrumentation enabled, they get orphan spans. This can be easily reproduced, so this PR fixes the issue.

## Background

There are two "modes" WCF can operate in: delayed instrumentation, and "simple" instrumentation. Delayed instrumentation was introduced in https://github.com/DataDog/dd-trace-dotnet/pull/1992 to avoid a problematic error scenario. Instead of instrumenting early across the whole "channel", we delayed instrumentation. A subsequent PR https://github.com/DataDog/dd-trace-dotnet/pull/5206 updated the instrumentation point to ensure the wcf span was active while `IDispatchMessageInspector` were running. 

This was somewhat complex, due to the convoluted and legacy way wcf uses async and threading. Basically, we use a weak table to hold the current span, and us the `MessageRpc` object as a key. Later on, we retrieve the scope and dispose it.

## Implementation details

The missing part was that we were not "restoring" the wcf scope for when the actual Invoker implementations executed. The process was:

- Update the tests and snapshots to:
  - Record _all_ spans and traces (it makes it easier to [understand the snapshots using](https://andrewlock.github.io/DatadogSpanSnapshotViewer/))
  - Create a span inside the "service" calls.
    - These _should_ have been children of the wcf server spans, but they were not. They were orphaned
- Update the "Invoker" implementations to grab the wcf spans, make it active, and then deactivate it afterwards
- Regenerate the snapshots, and confirm the custom spans are no longer orphaned 🎉 

Additionally, do a small amount of refactoring:
- "Fix" nullability in some places
- Extract the "restore wcf span" functionality to `WcfCommon` as it's used in multiple places

## Test coverage

Covered by existing, after the tweaks above. The snapshots change like this:

Before the fix, after adding the additional custom spans, `WSHttpBinding` traces:

image

After the fix, `WSHttpBinding` traces:

image

Before the fix, after adding the additional custom spans,  WebHttp traces:

image

After the fix, Web Http traces

image

Unfortunately, this ends up being somewhat unstable in regards to sorting, so I did some more refactoring to separate the root span for each separate httpcall. This should make sure we have more stable ordering of the spans using the built-in sort order

The result is we get separate traces for each API call instead:

image

## Other details

I wonder if we really need a weak table here? 🤔 seeing as we have an instrumentation point that explicitly disposes the span, shouldn't we just remove the entry at the same time?

### CODE REVIEW

File: tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs [hunk 1]
```diff
@@ -57,6 +53,8 @@ private static async Task<int> Invoke_ServerSyncAdd_Endpoints(CalculatorClient c
 
             try
             {
+                using var scope = SampleHelpers.CreateScope("ServerEmptyActionAdd");
+                calculator.DatadogScope = scope;
```
Comment (zacharycmontoya) [inline tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs:57]: Since you have some integration tests in place, can we try to remove this `DatadogScope` property? If it doesn't affect the results, we should remove it so we have less custom code

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/WcfCommon.cs [hunk 2]
```diff
@@ -216,6 +221,47 @@ internal class WcfCommon
 
             return null;
         }
+
+        public static CallTargetState ActivateScopeFromContext()
```
Comment (zacharycmontoya) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/WcfCommon.cs:225]: super small nit: It seems like this method has slightly too many responsibilities in that it's calling both `Tracer.Instance.ActivateSpan(Span)` and returning a `CallTargetState`. The only refactoring that comes to mind is to make this a `TryGetScopeFromContext` and return the fields used to populate the CallTargetState. And then from the caller you can activate the span and create the `CallTargetState`, but then you're making each callsite have to repeat multiple lines of code. I'm not sure if that's better than what you've done here, so no need to take my suggestion

**Review by zacharycmontoya — APPROVED** [review]
LGTM. As discussed, we have a corner case of WCF message headers overriding HTTP message headers. I think this PR is good as-is so we can tackle that as a follow-up PR if you'd like

File: tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs (hunk 1 above)
Comment (andrewlock) [reply tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs:57]: Unfortunately, if we remove this scope, it will make the snapshots flakey. The nature of how WCF works means that all the client spans are identical, which in turn means the spans end up with a non-deterministic order if we remove these. This is also why I split each individual operation into a separate span, instead of having the single over-arching span. 

FWIW, I'm not sure why we need to set the `DatadogScope` property - AFAICT it's not used, but as it was doing it before, I cargo-culted instead of making more changes 😄

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/WcfCommon.cs (hunk 2 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/WcfCommon.cs:225]: Yeah, I agree with you, but then also it's easy to screw up and forget to capture one or more parts that you need.. plus we can delay capturing the distributed context unless we need it (though we _should_ always hit the "activate" path in normal behaviour, so maybe not a big deal)

As you say, given every callsite would need to copy the exact same code to use the API, I'm not sure it's worth splitting out. We _could_ still capture the previous and then return a different carrier object instead of `CallTargetState` from the method, and then _convert_ it into a `CallTargetState` at the callsite, but that seems like more hassle for little gain? 🤷‍♂️ meh 😅

Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/WcfCommon.cs:225]: Agreed: meh 😆

File: tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs (hunk 1 above)
Comment (zacharycmontoya) [reply tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs:57]: Ah sorry, yes what I meant was removing the line `calculator.DatadogScope = scope;`. I understand that the manual scope in general helps with snapshot stability 👍🏼

Comment (andrewlock) [reply tracer/test/test-applications/integrations/Samples.Wcf/Client/Startup.cs:57]: Cool, I'll do that in the follow up PR that addresses other minor issues 🙂
